### PR TITLE
Add the ability to enable/disable registrations when in the OIDC flow

### LINKED ISSUE #11968: Allow disabling registration via OIDC and only limiting SSO to sign-in
**Description:**

For some setups that don't allow user registration it may be worthwhile to allow OIDC or OAuth2.0 sign-in for user convenience. However, I haven't found a way to disable creation of new users via OIDC. Sign-in works, but it also enables new users to sign-in to the server freely.

It would be great to change `allow_existing_users` (or have a new setting) with an enum:

- `no` = don't allow sign-in for an existing user
- `yes` = allow sign-in for a pre-existing user w/o OIDC mapping
- `only` = implies `yes`, but additionaly restricts the sign-in to match only pre-existing users

### PR DESCRIPTION
**Pull request by warrenbailey**
### Pull Request Checklist
Fixes #11968

Add a check for the enable registration flag in the OIDC flow. This allows the flow to limit SSO to sign in only, rather than automatically registering users that have a valid SSO login but do not have an account.

* [x] Pull request is based on the develop branch
* [x] Pull request includes a [changelog file](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#changelog). The entry should:
  - Be a short description of your change which makes sense to users. "Fixed a bug that prevented receiving messages from other servers." instead of "Moved X method from `EventStore` to `EventWorkerStore`.".
  - Use markdown where necessary, mostly for `code blocks`.
  - End with either a period (.) or an exclamation mark (!).
  - Start with a capital letter.
  - Feel free to credit yourself, by adding a sentence "Contributed by @github_username." or "Contributed by [Your Name]." to the end of the entry.
* [x] Pull request includes a [sign off](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#sign-off)
* [x] [Code style](https://matrix-org.github.io/synapse/latest/code_style.html) is correct
  (run the [linters](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#run-the-linters))

### CODE REVIEW

**Review by MatMaul — COMMENTED** [review]
I think we want a specific configuration toggle inside the OIDC configuration bloc for that.

Current behavior matches what is expected for most enterprise OIDC deployment: no open registration by specifying `enable_registration: false`, but possible to register through OIDC.

This would not be possible anymore with what you are proposing.

**Comment by warrenbailey** [comment]
Ok no worries. I can look at adding a specific configuration value and using that instead then.

**Comment by clokep** [comment]
Please shout when you think this is ready for another review.

**Comment by warrenbailey** [comment]
I will do, I'm just going to test it against my use case then should be good to go.

**Comment by villepeh** [comment]
Thank you for this. 

If this works out, I'd also consider my issue #13238 fixed by this, since the point of that issue ticket was to control registration through OIDC rather than creating allow-lists with (hacky-ish) `attribute_requirements`. 

However I understand some might still want that feature in particular and there's a PR for that, too: #12488, but it seems to be abandoned for now.

**Comment by warrenbailey** [comment]
@clokep I tested this on my server and it works for me. Ready for another review when you guys have a chance

**Comment by clokep** [comment]
> @clokep I tested this on my server and it works for me. Ready for another review when you guys have a chance

Great! We've add it back to the review queue. Thanks!

**Review by MatMaul — CHANGES_REQUESTED** [review]
Sorry for the delay. I think we are just missing a description of the configuration [here](https://github.com/matrix-org/synapse/blob/develop/docs/usage/configuration/config_documentation.md#oidc_providers) and we should be good to go !

File: changelog.d/14978.bugfix [hunk 1]
```diff
@@ -0,0 +1 @@
+Add the ability to enable/disable registrations when in the OIDC flow.
```
Comment (clokep) [inline changelog.d/14978.bugfix:1]: This sounds more like a feature than a bugfix to me?

Comment (warrenbailey) [reply changelog.d/14978.bugfix]: renamed 👍

**Review by MatMaul — APPROVED** [review]

**Comment by remram44** [comment]
The OIDC documentation should definitely be updated for this: https://matrix-org.github.io/synapse/latest/openid.html

**Comment by clokep** [comment]
> The OIDC documentation should definitely be updated for this: [matrix-org.github.io/synapse/latest/openid.html](https://matrix-org.github.io/synapse/latest/openid.html)

The configuration documentation was updated, see [here](https://matrix-org.github.io/synapse/latest/usage/configuration/config_documentation.html#oidc_providers). The page you linked to is mostly examples, it doesn't document every configuration option.

**Comment by remram44** [comment]
Maybe you should rename the heading in the left navigation then?

This is a configuration setting that is **specific to OpenID Connect**. It is really awkward that it is not documented in the doc section titled "OpenID Connect".

**Comment by clokep** [comment]
> This is a configuration setting that is **specific to OpenID Connect**. It is really awkward that it is not documented in the doc section titled "OpenID Connect".

The page [links back to the full docs](https://matrix-org.github.io/synapse/latest/openid.html#preparing-synapse):

> To enable the OpenID integration, you should then add a section to the `oidc_providers` setting in your configuration file. See the [configuration manual](https://matrix-org.github.io/synapse/latest/usage/configuration/config_documentation.html#oidc_providers) for some sample settings, as well as the text below for example configurations for specific providers.

We very much like pull requests that improve our documentation, however. It can be quite hard to know where people might look (or what search terms they might use) to find documentation!

**Pull request merged**